src/App.tsx

```diff
@@ -1,61 +1,135 @@
+);
+
+const getUserByName = (name: string): User | null => (
+  usersFromServer.find(user => user.name === name) || null
+);
+
+const refactoredTodos: Todo[] = todosFromServer.map((todo) => ({
+  ...todo,
+  user: getUserById(todo.userId),
+}));
 
 export const App = () => {
+  const [todos, setTodos] = useState(refactoredTodos);
+  const [title, setTitle] = useState('');
+  const [selectedUser, setSelectedUser] = useState('');
+  const [titleValid, setTitleValid] = useState(true);
+  const [userValid, setUserValid] = useState(true);
+
+  const handleTitleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setTitleValid(Boolean(event.target.value));
+  };
+
+  const handleUserChange = (event: ChangeEvent<HTMLSelectElement>) => {
+    setSelectedUser(event.target.value);
+    setUserValid(Boolean(event.target.value));
+  };
+
+  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    setTitleValid(Boolean(title));
+    setUserValid(Boolean(selectedUser));
+
+    if (!selectedUser || !title) {
+      return;
+    }
+
+    setTodos((currentTodos) => {
+      const maxId = Math.max(...currentTodos.map(todo => todo.id));
+      const newUser = getUserByName(selectedUser);
+
+      return [
+        ...currentTodos,
+        {
+          id: maxId + 1,
+          title,
+          completed: false,
+          userId: newUser ? newUser.id : null,
+          user: newUser,
+        },
+      ];
+    });
+
+    setTitle('');
+    setSelectedUser('');
+  };
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/users" method="POST">
+      <form
+        action="/api/users"
+        method="POST"
+        onSubmit={handleFormSubmit}
+      >
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <label>
+            {'Title: '}
+
+            <input
+              type="text"
+              data-cy="titleInput"
+              value={title}
+              onChange={handleTitleChange}
+            />
+            {!titleValid && (
+              <span className="error">Please enter a title</span>
+            )}
+          </label>
         </div>
 
         <div className="field">
-          <select data-cy="userSelect">
-            <option value="0" disabled>Choose a user</option>
-          </select>
+          <label>
+            {'User: '}
+
+            <select
+              data-cy="userSelect"
+              value={selectedUser}
+              onChange={handleUserChange}
+            >
+              <option
+                value=""
+                disabled
+              >
+                Choose a user
+              </option>
+
+              {usersFromServer.map((user) => (
+                <option
+                  key={user.id}
+                  value={user.name}
```

Comment:
Use `userId` as value in select.
That helps you to remove function `getUserByName` and will be more correct.